cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt

```diff
@@ -680,20 +681,101 @@ class ScopeManager : ScopeProvider {
      * @return a list of possible functions
      */
     @JvmOverloads
-    fun resolveFunction(
+    fun resolveFunctionLegacy(
         call: CallExpression,
         startScope: Scope? = currentScope
     ): List<FunctionDeclaration> {
         val (scope, name) = extractScope(call, startScope)
 
         val func =
             resolve<FunctionDeclaration>(scope) {
-                it.name.lastPartsMatch(name) && it.hasSignature(call)
+                it.name.lastPartsMatch(name) && it.hasSignature(call.signature, call.arguments)
             }
 
         return func
     }
 
+    class CallResolutionResult(
+        /** The original call expression. */
+        val call: CallExpression,
+        var candidateFunctions: List<FunctionDeclaration>,
+        var viableFunctions: Set<FunctionDeclaration>,
+        var signatureResults: Map<FunctionDeclaration, SignatureResult>,
+        var bestViable: List<FunctionDeclaration>,
+        var type: Type,
+        /**
+         * The actual start scope of the resolution, after [extractScope] is called on the callee.
+         * This can differ from the original start scope parameter handed to [resolveCall] if the
+         * callee contains an FQN.
+         */
+        var actualStartScope: Scope?
+    ) {
+
+        enum class Type {
+            OK,
```

Comment:
Doc at least for PROBLEMATIC (= a state that should never happen) and AMBIGUOUS (= can happen but there are multiple options e.g. due to overriding. What's the issue here that prevents us from saying "OK"?)

Reply:
Do you have a better naming idea for `Type`?

Reply:
I renamed it to `SuccessKind` for now